File: frontend/src/privacy.ts
// Narrow PAN/CVV detection; keep in sync with backend/privacy.py.
const cvv = /\b(?:cvv2?|cvc2?)\s*[:=№-]?\s*[0-9]{3,4}(?![0-9])|(?:код\s+безопасности|защитный\s+код|қауіпсіздік\s+коды)\s*[:=№-]?\s*[0-9]{3,4}(?![0-9])/iu;
const productCode = /(?:артикул|код\s+товара|sku|ean|gtin)\s*[:#№=-]?\s*$/iu;
const paymentContext = /(?:^|[^\p{L}\p{N}_])(?:карт[аыуые]|карточк\p{L}*|card|visa|mastercard|cvv2?|cvc2?|pan)(?![\p{L}\p{N}_])/iu;

function luhn(digits: string): boolean {
  if (new Set(digits).size === 1) return false;
  const sum = [...digits].reverse().reduce((total, digit, index) => {
    let value = Number(digit);
    if (index % 2) { value *= 2; if (value > 9) value -= 9; }
    return total + value;
  }, 0);
  return sum % 10 === 0;
}

export function containsPaymentData(text: string): boolean {
  if (cvv.test(text)) return true;
  const hasContext = paymentContext.test(text);
  const pan = /(?<![\p{L}\p{N}_-])[0-9](?:[ \t\u00a0-]*[0-9]){12,18}(?![\p{L}\p{N}_-])/gu;
  for (const match of text.matchAll(pan)) {
    const endings = [...match[0].matchAll(/[0-9](?=[ \t\u00a0-]|$)/g)].map(end => end.index + 1);
    const candidates = endings.map(end => match[0].slice(0, end).replace(/[^0-9]/g, ""));
    if (!candidates.some(digits => digits.length >= 13 && digits.length <= 19 && luhn(digits))) continue;
    if (!hasContext && productCode.test(text.slice(Math.max(0, match.index - 40), match.index))) continue;
    return true;
  }
  return false;
}

export const PAYMENT_REDACTED = "Сообщение с платёжными реквизитами не сохранено.";

File: backend/privacy.py
"""Reject recognisable payment credentials before retaining or forwarding text.

This is a narrow guard for PAN/CVV, not a general personal-data classifier.
Keep the browser storage guard in frontend/src/privacy.ts in sync.
"""
import re

from .errors import AppError

_PAN = re.compile(r'(?<![\w-])[0-9](?:[ \t\u00a0-]*[0-9]){12,18}(?![\w-])')
_CVV = re.compile(
    r'\b(?:cvv2?|cvc2?|код\s+безопасности|защитный\s+код|қауіпсіздік\s+коды)'
    r'\s*[:=№-]?\s*[0-9]{3,4}(?![0-9])', re.I)
_PRODUCT_CODE = re.compile(r'(?:артикул|код\s+товара|sku|ean|gtin)\s*[:#№=-]?\s*$', re.I)
_PAYMENT_CONTEXT = re.compile(r'\b(?:карт[аыуые]|карточк\w*|card|visa|mastercard|cvv2?|cvc2?|pan)\b', re.I)


def _luhn(digits):
    if len(set(digits)) == 1:
        return False
    total = 0
    for index, digit in enumerate(reversed(digits)):
        value = int(digit)
        if index % 2:
            value *= 2
            if value > 9:
                value -= 9
        total += value
    return total % 10 == 0


def contains_payment_data(text):
    if _CVV.search(text):
        return True
    payment_context = bool(_PAYMENT_CONTEXT.search(text))
    for match in _PAN.finditer(text):
        candidate = match.group()
        # Do not let a following expiry month/CVV joined by spaces hide a
        # complete PAN. Only stop at a separator, never inside a digit group.
        endings = [end.end() for end in re.finditer(r'[0-9](?=[ \t\u00a0-]|$)', candidate)]
        candidates = [re.sub(r'[^0-9]', '', candidate[:end]) for end in endings]
        if not any(13 <= len(digits) <= 19 and _luhn(digits) for digits in candidates):
            continue
        # An explicitly labelled catalogue code must remain searchable. A
        # payment context takes precedence over a misleading article label.
        if not payment_context and _PRODUCT_CODE.search(text[max(0, match.start()-40):match.start()]):
            continue
        return True
    return False


def reject_payment_data(text):
    if contains_payment_data(text):
        raise AppError('payment_data',
            'Не отправляйте номер банковской карты или код CVV/CVC. '
            'Удалите платёжные реквизиты и повторите запрос о товаре.', 422)
